### ISSUE TITLE
pkgconfig Requires.private and removal of static libraries

### ISSUE THREAD

**Issue by hmaarrfk** [conda-forge/conda-forge.github.io#1880]
### Your question:

There has been a trend at conda-forge to remove static libraries and to split off runtime libraries from build time dependencies in order to save download space.

There has also been a combined trend to add `pkgconfig` files to enable more packages that were historically built only for linux, be able to get built for windows.

However, the inclusion of the `Requires.private` flag in the pkgconfig files makes it necessary to add back build time dependencies in order to "help the static builds" (see the last FAQ question in https://people.freedesktop.org/~dbn/pkg-config-guide.html) 

I'm really not sure what to do about this. For example, many users have had to add back zlib as a host dependency (when their software doesn't link to it directly).

Second, the glib -> harfbuzz dependency is also a tough one to manage.

Thank you for your advice.

**Comment by hmaarrfk** [comment]
Taking `harfbuzz` as an example package Isn't this command useful?
```
pkg-config --cflags harfbuzz
```

Generating the output:
```
-I/home/mark/mambaforge/envs/dev/include/harfbuzz -I/home/mark/mambaforge/envs/dev/include/freetype2 -I/home/mark/mambaforge/envs/dev/include/libpng16 -I/home/mark/mambaforge/envs/dev/include -I/home/mark/mambaforge/envs/dev/include/glib-2.0 -I/home/mark/mambaforge/envs/dev/lib/glib-2.0/include -I/home/mark/mambaforge/envs/dev/include
```
is inherently useful to me.

**Comment by hmaarrfk** [comment]
The problem is that `pkgconfig` suggests that the file looks like:

```
prefix=/home/mark/mambaforge/envs/mcam_dev
includedir=${prefix}/include
libdir=${prefix}/lib

Name: harfbuzz
Description: HarfBuzz text shaping library
Version: 6.0.0
Requires.private: freetype2, graphite2, glib-2.0
Libs: -L${libdir} -lharfbuzz
Libs.private: -pthread -lm
Cflags: -I${includedir}/harfbuzz
```

whereas I think that for conda-forge, where we encourage dynamic linking, we should encourage the removal of the line `Requires.Private`.

Editing the file `harfbuzz.pc` so that it looks like
```
prefix=/home/mark/mambaforge/envs/mcam_dev
includedir=${prefix}/include
libdir=${prefix}/lib

Name: harfbuzz
Description: HarfBuzz text shaping library
Version: 6.0.0
# Requires.private: freetype2, graphite2, glib-2.0
Libs: -L${libdir} -lharfbuzz
Libs.private: -pthread -lm
Cflags: -I${includedir}/harfbuzz
```

makes the flags provided by `pkg-config` much more concise since the end user doesn't need to link to `freetype2` themselves:
```
pkg-config --cflags harfbuzz
-I/home/mark/mambaforge/envs/mcam_dev/include/harfbuzz
```

**Comment by carterbox** [comment]
Removal of static library is policy. https://github.com/conda-forge/cfep/blob/main/cfep-18.md

We already patch packages in order to devendor dependencies and ensure that artifacts are installed into the conda-environment correctly and avoid building static libs. Patching pc files to prevent static/unused linkages seems like a routine and reasonable thing.

There is probably an easy call to `sed` that can be added at the end of a build script to remove any lines beginning with `Requires.private` Something like:

```bash
sed -i '/Requires.private:/d' ./example.pc
```

**Comment by hmaarrfk** [comment]
3 challenges:

1. Windows equivalent command? It would be great if this could work without any additional dependencies (I know this is hard).
2. OSX Equivalent command? `sed -i` doesn't do the same thing on OSX.
3. There are some libraries where it was decided to explicitly go against that policy: https://github.com/conda-forge/zlib-feedstock/blob/main/recipe/meta.yaml#L105  https://github.com/conda-forge/zlib-feedstock/pull/64#issuecomment-1213353037 Zlib is only an example, but would be OK to break pkg-config in such situations?

**Comment by carterbox** [comment]
1.2. https://github.com/uutils/coreutils is promising, but they haven't implemented `sed` yet. Otherise, roll your own python script? 🤷🏼 😄 

3. zlib doesn't follow the policy "to the letter", but it does follow the spirit. "to the letter" would be putting the static libs in a `libzlib-static` package, but in the spirit of the policy, they are separated from the dynamic libraries in `zlib`. Downstream packages can link to `libzlib` without getting the static libs by accident.

**Comment by jakirkham** [comment]
Just following up on this question

> OSX Equivalent command? sed -i doesn't do the same thing on OSX.

Would use `sed -i.bak`. Basically `sed` on macOS won't do in-place writing without a backup file being written. `sed -i.bak` says add a `.bak` extension/suffix to the file in question and keep that as a backup. Can do this in the source directory (to keep these files from showing up in the package accidentally).

Would also note this behavior works fine on Linux. So can be used on all Unix-y platforms.

Lots of examples of this in conda-forge including [the build scripts of `python`]( https://github.com/conda-forge/python-feedstock/blob/21038d707640da7b5da152a5239e199c720f0574/recipe/build_base.sh#L463 ).

**Comment by pkgw** [comment]
I'm not sure what exact bug @isuruf is referring to, but I am not sure if removing the `Requires.private` lines is the right solution here. One area where these lines factors in is static vs. dynamic linking, but they also matter if you want to compile *any* kind of C/C++ code against a library.

Specifically, the example given above uses the `--cflags` option, which has nothing to do with linking — that option reports C/C++ compiler flags. And it would not be correct to truncate the output to just be `-I$PREFIX/include/harfbuzz`. The Harfbuzz headers reference types defined in the Glib/freetype/etc. headers and you simply can't (reliably) compile C/C++ code against Harfbuzz without including those headers as well. This is the same *category* of issue as you run into with static linking — the C/C++ development files for Harfbuzz have to depend on the C/C++ development files for Glib/freetype/etc, and so on throughout the dependency graph — but it's a different issue, and the fact that we don't encourage/support static linking doesn't make it go away, unfortunately. (And forcing people to write `pkg-config --cflags harfbuzz glib freetype2` induces the exact same problem, just in a more error-prone and less scalable way.) 

Ultimately, as long as we are compiling C/C++ code, the development files for libraries will have to depend on the development files for their dependents. (And I think that you can argue that this is really the essence of what it means to be a "library" regardless of language/OS/whatever, but that's beyond the scope here ...) The *real* issue here, in my view, is that there are still many cases where we don't (or don't properly) separate libraries and development files into different subpackages. You can't get rid of the dependency graph, but  you can at least do a better job of partitioning it into runtime/dev components. Unfortunately, Conda didn't include this distinction when it was first developed, so all of this separation has to be applied retroactively. (Which is a bit frustrating as Linux distros have been getting this right for 30 years, but there is also a reason that Conda has achieved the level of success that it has ...)

**Comment by hmaarrfk** [comment]
So in other words, resolve this issue correctly, we would need to really create `-dev` packages (or whatever name we choose) and ensure that the `-dev` package of the dependencies are included.


> but there is also a reason that Conda has achieved the level of success that it has ...

I think may be we should have a place where we can write why we chose to use conda(-forge). It is easy to forget sometimes.

**Comment by pkgw** [comment]
That's my belief, yes. But as @isuruf correctly points out, the current implementation of pkg-config does have bugs regarding this stuff, and there may be individual cases where it's correct to tweak the pkg-config file. But *if* the upstream developers have done things right, I think that in general the dependencies induced by `Requires.private` should be kept — they are basically developer dependencies that will need to be present to compile any kind of code against a library, not just for static linking purposes.

**Comment by traversaro** [comment]
> This is a bug in pkg-config, but `pkg-config` is not actively maintained anymore AFAIK. We should change to the new actively maintained implementation `pkgconf` and let that provide `pkg-config`
> 
> See also https://fedoraproject.org/wiki/Changes/pkgconf_as_system_pkg-config_implementation

As I experienced a problem related to this, I noticed that now a `pkgconf` packages is now available at https://github.com/conda-forge/pkgconf-feedstock (it was added in https://github.com/conda-forge/staged-recipes/pull/27048). Probably the next step is to provide some kind of `pkg-config` package that installs a symlink/redirect script to `pkgconf` ?

**Comment by h-vetinari** [comment]
> Probably the next step is to provide some kind of `pkg-config` package

We already have a [`pkg-config`](https://github.com/conda-forge/pkg-config-feedstock) package. Are you saying that we should switch our `pkg-config` to use `pkgconf` under the hood?

I could get behind that (seeing how `pkg-config` is [unmaintained](https://gitlab.freedesktop.org/pkg-config/pkg-config/-/tags) these days), though it's worth noting that `pkgconf` is ([intentionally](https://github.com/pkgconf/pkgconf#compatibility-with-pkg-config)) not bug-for-bug compatible with `pkg-config`.

PS. To complete the naming confusion, that there's [also](https://conda-metadata-app.streamlit.app/?q=conda-forge%2Flinux-64%2Fpkgconfig-1.5.5-py39hf3d152e_3.tar.bz2) a `pkgconfig` python(!) package. No idea what it's equivalent to or with, but I'm not advocating that we use it anyway - better to keep python out of the build environments where it's not necessary.

**Comment by hmaarrfk** [comment]
> Are you saying that we should switch our pkg-config to use pkgconf under the hood?

I somewhat prefer we don't do too many unsanctionned things like this. 

I believe we should experiment with a deliberate switch over using linter warnings when are are confident with the new strategy.

**Comment by isuruf** [comment]
> I somewhat prefer we don't do too many unsanctionned things like this.

Other distribtutions like debian, fedora do this, so it's not unheard of nor untested.

**Comment by hmaarrfk** [comment]
@isuruf one development since our conversation started is that
pkgconf now honors the `Requires.private` flags as it is done in `pkg-config`.

https://github.com/pkgconf/pkgconf/pull/353

Thoughts?

**Comment by pkgw** [comment]
I took some time to revisit this and related issues this evening. I really wanted to be able to say "You know what, let's just go ahead and strip these things out", but as I remind myself of how these build tools work, I'm still convinced that removing these lines is simply **not correct**, and has the potential to break our builds in ways both obvious and subtle.

If the private requirements were only used for static library linking, I'd be fine with removing them. But that's **not the case**.

The private requirements are also used for the generation of C compiler flags when you're compiling code against a library, and as a general matter *you simply cannot safely omit the cflags of dependent libraries when you do that*. It *might* work to omit these flags in individual instances, but it's not something that we can rely on in the general case. And if it works at one point in time, there's no way to be sure that it will continue working in the future.

The most obvious failure mode is that library A depends on library B, and library A's headers include library B's headers, and those headers are stored in a place that's not directly in the default search path (i.e. `$PREFIX/include` for us). Removing B as a private dep removes that extra search path and breaks any builds that don't happen to bring in B as a dep in some other way. The streamlining of the cflags output mentioned in https://github.com/conda-forge/conda-forge.github.io/issues/1880#issuecomment-1384352376 might be convenient, and it might not break the build in this particular example, but it is wrong.

The more insidious failure mode has to do with preprocessor definitions. These can and do occur in Cflags lines — here are a few examples my Fedora system: 

```
    /usr/lib64/pkgconfig/avahi-core.pc: Cflags: -D_REENTRANT -I${includedir}
          /usr/lib64/pkgconfig/fuse.pc: Cflags: -I${includedir}/fuse -D_FILE_OFFSET_BITS=64
/usr/lib64/pkgconfig/libpcre2-posix.pc: Cflags: -I${includedir} -DPCRE2POSIX_SHARED
          /usr/lib64/pkgconfig/zlib.pc: Cflags: -I${includedir} -DWITH_GZFILEOP
```

(There are plenty of examples of this in conda-forge pkg-config files as well.) If you remove the private requirements, you stop getting these definitions. By definition, they alter the interpretation of the dependent library's headers, and usually they're in the pkg-config file precisely because they control low-level stuff like byte widths, endianness, and feature availability. Missing settings like this might lead to compile failures ... or they might lead to code that compiles successfully but does not actually work correctly at runtime. And because we're talking about C, the implications of the incorrectness could range from minor issues to severe security problems — there's just no way to know.

I understand that it is very annoying and non-obvious to need to manually pull in additional dev dependencies just to get pkg-config to verify that a library is available, but unfortunately most people and scripts do that using `pkg-config --cflags`, which is exactly the time that you need to pull all of the info from the dependents.

I'll repeat that if the only issue were static libraries, this wouldn't be a problem, but we need to be able to compile against our own libraries. We might be able to get away with removing the private deps in some, or maybe even most, situations, but it is basically adding intentional incorrectness to the way that we compile low-level, security-sensitive code, which is not something that I'm comfortable with.

**Comment by traversaro** [comment]
> The most obvious failure mode is that library A depends on library B, and library A's headers include library B's headers, and those headers are stored in a place that's not directly in the default search path (i.e. $PREFIX/include for us). 

Unless I am missing something, if "library A's headers include library B's headers" then `library B` should be in library A's `Requires` list, not `Requires.private`.

To double check this, I looked in https://www.freedesktop.org/wiki/Software/pkg-config/ that links https://people.freedesktop.org/~dbn/pkg-config-guide.html, that seems non-ambiguous on this, see the following quotes:

> Requires.private: A list of private packages required by this package but not exposed to applications. The version specific rules from the Requires field also apply here.

> Requires and Requires.private define other modules needed by the library. It is usually preferred to use the private variant of Requires to avoid exposing unnecessary libraries to the program that is linking with your library. If the program will not be using the symbols of the required library, it should not be linking directly to that library. 

> Since pkg-config always exposes the link flags of the Requires libraries, these modules will become direct dependencies of the program. **On the other hand, libraries from Requires.private will only be included when static linking.** For this reason, it is usually only appropriate to add modules from the same package in Requires. 

In particular, the last quote seems to indicate that `Requires.private` is indeed related only to static linking.

**Comment by pkgw** [comment]
@traversaro That documentation is correct as it applies to linking, but for better or for worse, private requires are specified to appear in cflags lines as well. See discussion here:

https://github.com/pkgconf/pkgconf/issues/300

I don't fully understand the motivation for the design (quite possibly a bug), but the framework has been around for long enough that it would be a breaking change to alter it.

**Comment by hmaarrfk** [comment]
The audit you made is cool. 

I think there are some C packages that use preprocessor macros to control features like that. 

I’m of the opinion that this bad design but I know it is unavoidable when you are a consumer of packages. 

I would prefer to consider these on a case by case basis. And rather raise the issues upstream. 

If you have your grep command. I would love to run it on my environment.

**Comment by isuruf** [comment]
For the case of libglib, I think we should have `libglib-devel` and move the headers and the pc files there. Then we can add `libzlib-devel` as a requirement to `libglib-devel`.

**Comment by pkgw** [comment]
My grep command was not very fancy 😄 : `grep 'Cflags.*-D' /usr/lib64/pkgconfig/*.pc`. There might be other "interesting" flags that it fails to pick up, but preprocessor definitions are the ones that come to my mind first as risky to omit.

Adding a `libglib-devel` should hopefully be a bit easier since I recently migrated the recipe to V1 / rattler-build, now that it supports "staging builds". I'm finding it easier to work with for multi-output builds, at least.

Also, @hmaarrfk I think in some of the issues related to this topic you've mentioned cases where the private requires were causing problems that seemed harder to resolve than "add a few unexpected dependencies". I will be happy to take a look and try to help get those issues straightened out.

We could also consider looking into patching our pkg-config to try to give a clearer message when package A is "not found" because one of its dependent `.pc` files is missing. That might not be too hard to do?

**Comment by hmaarrfk** [comment]
> seemed harder to resolve than "add a few unexpected dependencies".

This is true for you and I.
However my strategy is expensive in terms of my time:

1. Get into `conda debug`
2. Execute `pkconf --exists library`
3. Go to conda-forge.org/packages
4. Install a package that looks related
5. Repeat step 2.

This is super tiring.

-----------------------------------------

```
find ${CONDA_PREFIX} -name '*.pc' -exec grep -H 'Cflags.*-D' {} +
```
seems to be the command for your conda environment.

I'm not seeing anything that would be a "transitive dependency".

-----------------------------------------

Basically, in my mind, if you have something like

libA and libB, and libB's header contains

```
#if LIBA_FEATURE_FLAG
change the behavior of libB without the inclusion of #include <libA_header.h>
#endif
```

this seems like a bug in libB.

There seems to be some scary macros like `-DNOMINMAX` that have that kind of effect.

I would somewhat consider
https://github.com/zlib-ng/zlib-ng/blob/c69e75e9e24bba709752268df239fee3c2e3dcfa/zconf-ng.h.in#L154

for the zlib library `WITH_GZFILEOP` to be a bug. and I would advocate to patch it out.

**Comment by kleisauke** [comment]
> This is a bug in pkg-config, but `pkg-config` is not actively maintained anymore AFAIK. We should change to the new actively maintained implementation `pkgconf` and let that provide `pkg-config`
> 
> See also https://fedoraproject.org/wiki/Changes/pkgconf_as_system_pkg-config_implementation

+1. Switching to `pkgconf` would resolve the issue on Windows where `Libs.private` dependencies are always included. In `pkg-config`, this behavior is hard-coded[^1]:
https://gitlab.freedesktop.org/pkg-config/pkg-config/-/blob/pkg-config-0.29.2/config.h.win32.in#L6-7

whereas `pkgconf` fixed this in https://github.com/pkgconf/pkgconf/pull/410.

See e.g. issue https://github.com/conda-forge/libxml2-feedstock/issues/164 and PR https://github.com/conda-forge/pyvips-feedstock/pull/40 where this is causing issues.

[^1]: While it can be disabled with `--disable-indirect-deps` when building via Autotools, the `pkg-config` feedstock is built using NMake Makefiles, so that option is not available there.
